projects/hathor-wallet-headless/0004-tx-transaction.md

```diff
@@ -0,0 +1,209 @@
+# Summary
+[summary]: #summary
+
+Create a transaction template that can be used to form a transaction by following the instructions.
+
+# Motivation
+[motivation]: #motivation
+
+Some clients require special additions to how the transactions are created (e.g. sending authority to certain addresses after the transaction, allowing the authority address to be from outside the wallet, adding a special output in the transaction, etc.).
+Each of these would be added as an option on the wallet facade and allowing the wallet-lib to create the desired transaction, but each option requires discussion, development time and a new release from both the headless and wallet-lib.
+
+A solution to this is a more flexible API that can receive a transaction template and follow its intructions to form a transaction, if the template is well contructed it can solve all special cases without additional development.

```

Comment:
thought: I understand this template approach adds a new set of possibilities, but it also has some overlaps with our existing `tx_proposal` structure already present at the headless wallet:
- https://github.com/HathorNetwork/hathor-wallet/issues/308
- https://github.com/HathorNetwork/hathor-wallet-headless/pull/211

Could you please elaborate a bit on how they are meant for different purposes or maybe even explicit their overlap?

Reply:
The transaction proposal for atomic swap allows partial construction of a transaction so multiple participants can take turns adding their inputs and outputs.
The template is meant for a single wallet (but could be expanded in the future) and the objective is to build transactions with special cases to avoid adding more flags to the wallet facade everytime.

One of the examples i gave is a mint transaction that sends the created mint authority to an address outside the wallet that on the wallet facade required 2 options, one to specify the destination address of the authority and another to disable the check that this address is not from our wallet.
If we were using the template this would be very straightforward, just specify the destination address on the authority output.
Since the template is a "special cases" api we allow the user more freedom, specially from our security constraints.

Reply:
It makes sense. I guess that just by expliciting the purpose of each tool in the future docs will be enough to avoid any ambiguity here.